application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java

```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
         Assert.assertEquals(0, pageData.getData().size());
     }
 
+    @Test
+    public void testDeleteDashboardExistsRelation_Error_RestoreRelation_DeleteRelation_DeleteDashboard_Ok() throws Exception {
+        Dashboard dashboard = new Dashboard();
+        dashboard.setTitle("My dashboard");
+        Dashboard savedDashboard = doPost("/api/dashboard", dashboard, Dashboard.class);
+
+        Alarm savedAlarm = createAlarm(savedDashboard.getId());
+
+        AlarmOperationResult alarmOperationResult = alarmService.createOrUpdateAlarm(savedAlarm);
+        Assert.assertEquals("List of propagatedEntities is not equal to number of created propagatedEntities!",
+                1, alarmOperationResult.getPropagatedEntitiesList().size());
+        Assert.assertEquals("DashboardId in propagatedEntities is not equal savedDashboardId!",
+                savedDashboard.getId(), alarmOperationResult.getPropagatedEntitiesList().get(0));
+
+        DeviceProfile deviceProfile =  createDeviceProfile();
+        deviceProfile.setDefaultDashboardId(savedDashboard.getId());
+        DeviceProfile savedDeviceProfile = doPost("/api/deviceProfile", createDeviceProfile(), DeviceProfile.class);
+
+        savedDeviceProfile.setDefaultDashboardId(savedDashboard.getId());
+        DeviceProfile updatedDeviceProfile = doPost("/api/deviceProfile", savedDeviceProfile, DeviceProfile.class);
+
+        String typeRelation = EntityRelation.CONTAINS_TYPE;
```

Comment:
I suggest testing the corner cases
1. no relations (probably already tested earlier)
2. one relation (your test)
3. many relations (two or more relations). from Tenant to Dashboard, with different Types or any you want

Reply:
Ok